packages/js/product-editor/src/blocks/product-fields/downloads/edit.tsx

```diff
@@ -47,104 +46,50 @@ export function Edit( {
 		'product',
 		'downloads'
 	);
-	const [ fileItems, setFileItems ] = useState< DownloadableFileItem[] >(
-		[]
-	);
 
 	const { allowedMimeTypes } = useSelect( ( select ) => {
 		const { getEditorSettings } = select( 'core/editor' );
 		return getEditorSettings();
 	} );
 
-	useEffect( () => {
-		setFileItems( ( currentItems ) => {
-			const downloadsMap = downloads.reduce<
-				Record< string, ProductDownload >
-			>(
-				( current, download ) => ( {
-					...current,
-					[ download.id ]: download,
-				} ),
-				{}
-			);
+	const allowedTypes = allowedMimeTypes
+		? Object.values( allowedMimeTypes )
+		: [];
 
-			const fileItemsInDownloads = currentItems.reduce<
-				DownloadableFileItem[]
-			>( function keepPresentDownload( items, item ) {
-				if ( item.download.id === '' ) {
-					items.push( item );
-					return items;
-				}
-				if ( item.download.id && item.download.id in downloadsMap ) {
-					const download = downloadsMap[ item.download.id ];
-					delete downloadsMap[ item.download.id ];
-					items.push( {
-						...item,
-						download,
-					} );
-					return items;
-				}
-				return items;
-			}, [] );
-
-			return Object.values( downloadsMap ).reduce<
-				DownloadableFileItem[]
-			>( function addAbsentDownload( items, download ) {
-				items.push( {
-					key: String( download.id ),
-					download,
-				} );
-				return items;
-			}, fileItemsInDownloads );
-		} );
-	}, [ downloads ] );
+	const { createErrorNotice } = useDispatch( 'core/notices' );
 
 	function handleFileUpload( files: MediaItem | MediaItem[] ) {
 		if ( ! Array.isArray( files ) ) return;
 
-		const { uploadedFiles, items } = files.reduce< {
-			uploadedFiles: Product[ 'downloads' ];
-			items: DownloadableFileItem[];
-		} >(
-			( current, file, index ) => {
-				const download = {
-					id: file.id ? String( file.id ) : '',
-					file: file.url,
-					name: getFileName( file.url ),
-				};
-				const item = {
-					key: `_${ index }`,
-					download,
-					uploading: ! Boolean( file.id ),
-				};
-
-				if ( download.id ) {
-					current.uploadedFiles.push( download );
-				}
-				current.items.push( item );
-
-				return current;
-			},
-			{ uploadedFiles: [], items: [] }
-		);
+		const uploadedFiles = files
+			.filter(
+				( file ) =>
+					! file.id ||
+					! downloads.some(
+						( download ) =>
+							download.id === String( file.id ) ||
+							download.file === file.url
+					)
+			)
+			.map( ( file ) => ( {
+				id: file.id ? String( file.id ) : '',
```

Comment:
Why are you stringifying this?
I get an explicit error back from the server `downloads[0][id] is not of type string.`
Which may also be an issue with the `Insert from URL`

Reply:
Well in fact, the server does required the download ids to be sent as string (even when the returned download object do have integer and UUID types for its id). Also the `Insert from URL` does not have and id that's why an empty string is being sent to the server, at server side a UUID is automatically generated and assigned to the download object.

If you see that the files are not being saved at the server then you should check the type of the product, currently `variable` products are not allowed to have downloads or be virtual. Please try the same test in a simple product.

Reply:
It seems to be fixed now.